nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/bigquery/PutBigQuery.java

```diff
@@ -0,0 +1,457 @@
+        CreateWriteStreamRequest createWriteStreamRequest = CreateWriteStreamRequest.newBuilder()
+            .setParent(tableName.toString())
+            .setWriteStream(WriteStream.newBuilder().setType(type).build())
+            .build();
+
+        return writeClient.createWriteStream(createWriteStreamRequest);
+    }
+
+    protected BigQueryWriteClient createWriteClient(GoogleCredentials credentials) {
+        BigQueryWriteClient client;
+        try {
+            client = BigQueryWriteClient.create(BigQueryWriteSettings.newBuilder().setCredentialsProvider(FixedCredentialsProvider.create(credentials)).build());
+        } catch (Exception e) {
+            throw new ProcessException("Failed to create Big Query Write Client for writing due to", e);
+        }
+
+        return client;
+    }
+
+    protected StreamWriter createStreamWriter(String streamName, Descriptors.Descriptor descriptor, GoogleCredentials credentials) throws IOException {
+        ProtoSchema protoSchema = ProtoSchemaConverter.convert(descriptor);
+        return StreamWriter.newBuilder(streamName)
+            .setWriterSchema(protoSchema)
+            .setCredentialsProvider(FixedCredentialsProvider.create(credentials)).build();
+    }
+
+    private boolean isBatch() {
+        return BATCH_TYPE.getValue().equals(transferType);
+    }
+
+    private static class AppendContext {
+        private final ProtoRows data;
+        private final long offset;
+        private int retryCount;
+
+        AppendContext(ProtoRows data, long offset) {
+            this.data = data;
+            this.offset = offset;
+            this.retryCount = 0;
+        }
+
+        public ProtoRows getData() {
+            return data;
+        }
+
+        public int getRetryCount() {
+            return retryCount;
+        }
+
+        public void incrementRetryCount() {
+            retryCount++;
+        }
+
+        public long getOffset() {
+            return offset;
+        }
+    }
+
+    private static Map<String, Object> convertMapRecord(Map<String, Object> map) {
+        Map<String, Object> result = new HashMap<>();
+        for (String key : map.keySet()) {
+            Object obj = map.get(key);
+            if (obj instanceof MapRecord) {
+                result.put(key, convertMapRecord(((MapRecord) obj).toMap()));
+            } else if (obj instanceof Object[]
+                && ((Object[]) obj).length > 0
+                && ((Object[]) obj)[0] instanceof MapRecord) {
+                List<Map<String, Object>> lmapr = new ArrayList<>();
+                for (Object mapr : ((Object[]) obj)) {
+                    lmapr.add(convertMapRecord(((MapRecord) mapr).toMap()));
+                }
+                result.put(key, lmapr);
+            } else if (obj instanceof Timestamp) {
+                // ZoneOffset.UTC time zone is necessary due to implicit time zone conversion in Record Readers from
+                // the local system time zone to the GMT time zone
```

Comment:
This should no longer be happening, are you sure this is the case?

Reply:
When I was working on the implementation I was struggling a lot Time / Timestamp it seemed that I'm always off with 1 or 2 hours when the data landed on the BigQuery side. I remember looking into what is happening and the RecordReader works with DateTypeUtils class. When the date/time is a long number (epoch) we are good however if it is a format string it uses the dateFormat to parse it, however for example the AbstractCsvRecordReader initialises the DateFormat as follwos
```
LAZY_TIME_FORMAT = () -> DataTypeUtils.getDateFormat(timeFormat);
```
and the DataTypeUtils.getDateFormat(timeFormat) is using the default timeZone "return getDateFormat(pattern, TimeZone.getDefault());" and I think that is where things go south but maybe I'm missing something I can look into this more deeply if needed.

Reply:
Thanks for the response. Not wanting to hold things up, but I think this is worth a bit more investigation. The `TimeZone.getDefault()` method returns the time zone according to the operating system and Java platform configuration, which is usually a local time zone, but could be UTC depending on the configuration. Timestamp strings without a zone offset should be handled using the default timezone. For that reason, performing this type of conversion in the processor itself should be avoided.

When timestamps are formatted as strings, they should include a timezone offset to guarantee the same behavior regardless of the system configuration. In absence of the timezone offset, it is best to fallback to the system default timezone, instead of attempting to make adjustments.

In older versions of NiFi, the DateTypeUtils had an internal setting that defaulted to GMT, but this created other complex problems in referencing components, so that behavior was moved.

With that background, the best approach is to avoid any automatic timezone conversion.

Reply:
I looked into the whole thing, tested it manually and adjusted the input and the integration tests as well. It looks like I managed to make it work and verify the correctness. Date and time handling is always a challenge in this case when RecordReader, NiFi execution environment and Google Big Query are all involved with mapping between different types along the way (java, grpc, google bigquery storage...) it is not trivial for sure :) Thanks for looking into it I had challenges with keeping my test in sync with the existing ones but I think the existing was not working correctly so removing it made everything clearer.

Reply:
Thanks for making the adjustments to the date and time handling. I know some older components previously included workarounds related to DateTypeUtils, so this was a good opportunity to streamline the approach. The updates look good.